package.ily

```diff
@@ -36,13 +36,21 @@
 
 %%%% export music
 % filebase: file basename - suffix (.krn/.xml) is taken from the exporter
-% exporter: symbol or function: hum -> humdrum, xml -> musicXML, not implemented yet: [l]mei -> [L-]MEI, lily -> LilyPond
+% exporter: symbol or function: hum -> humdrum, xml -> musicXML,
+%           not implemented yet: [l]mei -> [L-]MEI, lily -> LilyPond
 %           or an exporter function #(lambda (export-tree filename . options) ...)
 % music: the music to export
 #(define (symbol-or-procedure? v) (or (symbol? v)(procedure? v)))
 exportMusic =
-#(let ((exporters `((xml . ,exportMusicXML)(hum . ,exportHumdrum)(lily . ,exportLilyPond))))
-   (define-void-function (filebase exporter music)((string? (ly:parser-output-name)) symbol-or-procedure? ly:music?)
-     (if (symbol? exporter) (set! exporter (ly:assoc-get exporter exporters exportMusicXML #t)))
-     (ly:run-translator (ly:score-music (scorify-music music)) (FileExport `((filebase . ,filebase)(exporter . ,exporter)) ))
-     ))
+#(let
+  ((exporters
```

Comment:
@jpvoigt I have just started looking into the code, but I have a suggestion here. I think the current approach to provide and register the export functions is tedious, error-prone and not very "plugin-like". I'd replace this explicit alist with something like `((exporters (export-functions)))` with

* `export-functions` being defined in `api.scm`
* moving the export modules (`Humdrum.scm` etc.) to a subdirectory (changing the module path from `lilypond-export Humdrum` to, say, `lilypond-export exporters Humdrum`)
* having `api.scm` load *all* modules it can find in the subdirectory
* require an export module to export a function with a *generic* name (`export` instead of `exportMusicXML`)
* have `export-functions` create the alist from the `export` functions in all found export modules.

This way adding a new export target would not require adding information in three different places but only to create a new module with that one required function and drop that into the exporters directory.

What do you think?

Reply:
Yes, this is a good idea. This part is still from the code for my presentation at the MEI conference in Tours last year. We should go this way, but talk with Alex and Paul before that.

Reply:
We should do it like that, but talk with Alex and Paul before that.
Thanks for the cleanup!

Reply:
After starting to read into the MusicXML module and seeing how massive the export routines will become when approaching anything like comprehensive coverage I suggest yet another approach.

The current code (and my previous suggestion) first load *all* registered export modules and then select the appropriate function. Instead the process should be the following:

* check if `exporter` is a symbol or a procedure
* if it is a procedure simply use that (maybe make that more explicit)
* if it is a symbol then to something like `set!/let exporter (export-function exporter)`
* where `export-function` (defined in `api.scm`) loads the corresponding module and returns the generically named function. The "exporter" key then has to match the module name exactly.

Reply:
@PaulMorris  @rshura what do you think?

Reply:
I'll create a PR then it'll be easier to assess what I think of.